sdk/python/feast/registry.py

```diff
@@ -490,3 +495,72 @@ def _write_registry(self, registry_proto: RegistryProto):
         file_obj.seek(0)
         blob.upload_from_file(file_obj)
         return
+
+
+class S3RegistryStore(RegistryStore):
+    def __init__(self, uri: str):
+        self._uri = urlparse(uri)
+        self._bucket = self._uri.hostname
+        self._key = self._uri.path.lstrip("/")
+        return
+
+    def get_registry_proto(self):
+        import boto3
+        import botocore
+
+        file_obj = TemporaryFile()
+        registry_proto = RegistryProto()
+        s3 = boto3.resource("s3", endpoint_url=os.environ.get("FEAST_S3_ENDPOINT_URL"))
+        try:
+            bucket = s3.Bucket(self._bucket)
+            s3.meta.client.head_bucket(Bucket=bucket.name)
+        except botocore.client.ClientError as e:
+            # If a client error is thrown, then check that it was a 404 error.
+            # If it was a 404 error, then the bucket does not exist.
+            error_code = int(e.response["Error"]["Code"])
+            if error_code == 404:
+                raise FeatureBucketNotExist(self._bucket)
+            else:
+                raise FeatureBucketForbiddenAccess(self._bucket)
+
+        try:
+            obj = bucket.Object(self._key)
+            obj.download_fileobj(file_obj)
+            file_obj.seek(0)
+            registry_proto.ParseFromString(file_obj.read())
+            return registry_proto
+        except botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "404":
+                raise FileNotFoundError(
+                    f'Registry not found at path "{self._uri.geturl()}". Have you run "feast apply"?'
+                )
+            else:
+                raise FileNotFoundError(
+                    f'Registry is not able to locate data under path "{self._uri.geturl()}" with [original error]: {e.response}'
+                )
+
+    def update_registry_proto(
+        self, updater: Optional[Callable[[RegistryProto], RegistryProto]] = None
+    ):
+        try:
+            registry_proto = self.get_registry_proto()
+        except FileNotFoundError:
+            registry_proto = RegistryProto()
+            registry_proto.registry_schema_version = REGISTRY_SCHEMA_VERSION
+        if updater:
+            registry_proto = updater(registry_proto)
+        self._write_registry(registry_proto)
+        return
+
+    def _write_registry(self, registry_proto: RegistryProto):
+        import boto3
+
+        registry_proto.version_id = str(uuid.uuid4())
+        registry_proto.last_updated.FromDatetime(datetime.utcnow())
+        # we have already checked the bucket exists so no need to do it again
+        file_obj = TemporaryFile()
+        file_obj.write(registry_proto.SerializeToString())
+        file_obj.seek(0)
+        s3 = boto3.resource("s3", endpoint_url=os.environ.get("FEAST_S3_ENDPOINT_URL"))
+        s3.put_object(Bucket=self._bucket, Body=file_obj, Key=self._key)
```

Comment:
```suggestion
        self.s3.put_object(Bucket=self._bucket, Body=file_obj, Key=self._key)
```

Reply:
@leonid133 does:
```
s3.put_object(Bucket=self._bucket, Body=file_obj, Key=self._key)
```
work for you?
The code is using the `resource` API and not the `client` API so it fails with this error:
```
self.s3.put_object(Bucket=self._bucket, Body=file_obj, Key=self._key)
AttributeError: 's3.ServiceResource' object has no attribute 'put_object'
```

I think the syntax should be:
```
self.s3.Bucket(self._bucket).put_object(Body=file_obj, Key=self._key)
```

or else use the `client` API